hooks/image/60-kolibri-content

```diff
@@ -45,42 +66,122 @@ import_kolibri_channel()
-# At some point we should revisit this to try to make the caching work properly.
-rm -rf "${KOLIBRI_CONTENT_DIR}"
-mkdir -p "${KOLIBRI_CONTENT_DIR}"
-
-venv_dir="${EIB_TMPDIR}/kolibri-content-venv"
-python3 -m venv ${venv_dir}
-source ${venv_dir}/bin/activate
-
-pip install kolibri==${EIB_KOLIBRI_APP_VERSION}
-pip install kolibri-app-desktop-xdg-plugin==${EIB_KOLIBRI_APP_DESKTOP_XDG_PLUGIN_VERSION}
-pip install kolibri-desktop-auth-plugin==${EIB_KOLIBRI_DESKTOP_AUTH_PLUGIN_VERSION}
-
-export KOLIBRI_HOME="${KOLIBRI_CONTENT_DIR}"
-
-kolibri plugin enable kolibri.plugins.app
-kolibri plugin enable kolibri_app_desktop_xdg_plugin
-kolibri plugin enable kolibri_desktop_auth_plugin
-
-for channel_id in ${EIB_KOLIBRI_INSTALL_CHANNELS}; do
-  import_kolibri_channel "${channel_id}"
-done
-
-# Sort channels in the same order as in EIB_KOLIBRI_INSTALL_CHANNELS
-position=1
-for channel_id in ${EIB_KOLIBRI_INSTALL_CHANNELS}; do
-  kolibri manage setchannelposition ${channel_id} ${position} || true
-  let position=position+1
-done
-
-# Empty the user database, and ensure that each instance of this image has a
-# unique Facility ID.
-# <https://kolibri.readthedocs.io/en/latest/install/provision.html#prepare-the-kolibri-folder-for-copying>
-(echo yes; echo yes) | kolibri manage deprovision
+prepare_kolibri_env()
+{
+  local app_name=$1
+  # Needs to be kept in sync with hooks/image/61-kolibri-content-install
+  KOLIBRI_CONTENT_DIR="${EIB_CONTENTDIR}/${app_name}-content"
+  # FIXME: For now, we need to remove old content that may exist from previous
+  # runs to prevent images accidentally getting extra channels that are not in
+  # their configuration. However, this defeats the purpose of the image builder
+  # saving the contents of EIB_CONTENTDIR between runs to improve its performance.
+  # At some point we should revisit this to try to make the caching work properly.
+  rm -rf "${KOLIBRI_CONTENT_DIR}"
+  mkdir -p "${KOLIBRI_CONTENT_DIR}"
+
+  venv_dir="${EIB_TMPDIR}/${app_name}-content-venv"
+  python3 -m venv ${venv_dir}
+  source ${venv_dir}/bin/activate
+
+  pip install kolibri==${EIB_KOLIBRI_APP_VERSION}
+
+  export KOLIBRI_HOME="${KOLIBRI_CONTENT_DIR}"
+
+  if [ ${app_name} == "kolibri" ]; then
+    PLUGINS_DISABLE=()
+    PLUGINS_ENABLE=(
+      "kolibri.plugins.app"
+      "kolibri_app_desktop_xdg_plugin"
+      "kolibri_desktop_auth_plugin"
+    )
+    pip install kolibri-app-desktop-xdg-plugin==${EIB_KOLIBRI_APP_DESKTOP_XDG_PLUGIN_VERSION}
+    pip install kolibri-desktop-auth-plugin==${EIB_KOLIBRI_DESKTOP_AUTH_PLUGIN_VERSION}
+  elif [ ${app_name} == "endless-key" ]; then
+    PLUGINS_DISABLE=(
+      "kolibri.plugins.default_theme"
+      "kolibri.plugins.learn"
+    )
+    PLUGINS_ENABLE=(
+      "kolibri.plugins.app"
+      "kolibri_endless_key_theme"
+      # FIXME: enabling kolibri-explore-plugin fails because we are using a
+      #        kolibri version that is too old.
+      #"kolibri_explore_plugin"
```

Comment:
I think this is fine. It means Kolibri will need to execute some of extra database migrations on first run, but shouldn't be a problem.

One issue we are going to crash into here is that the Explore plugin provides its own [`applyexternaltags`](https://github.com/endlessm/kolibri-explore-plugin/blob/master/kolibri_explore_plugin/management/commands/applyexternaltags.py) management command. We use that when importing content from our content collection files, and it's necessary so the Explore plugin knows how to populate the different sections in the Discovery page. Without those tags, the Discovery page is pretty bare.

I think that is going to be a bit of a puzzle to sort out, so I don't know if we want to do it _here_, but we'll want to do this all a bit differently in the near future. The good news is, I don't think we will need to put a Kolibri database in the image at all any more, so this issue about the Explore plugin and the old Kolibri version will probably be moot. All we need is to fetch the content files, put them somewhere (as in `kolibri manage exportcontent`), add a JSON file describing what that content is (the `exportcontent` command generates a content manifest file), and get the app to populate its database (with the correct content in the database and marked available) based on that file.

For Kolibri, that JSON file would be exactly what the `exportcontent` command generates. For Endless Key, we'd copy our own JSON files from [endless-key-collections](https://github.com/endlessm/endless-key-collections). Those ones include the information we need about external content tags.

Reply:
It's certainly a lot simpler to put the content directly in the Kolibri homedir rather than putting it somewhere else on disk and then coming up with a procedure to import it and then maybe delete the massive duplicated data.

When I was previously thinking about this, my understanding was that the content pack would still be picked at runtime. Then the explore plugin would very quickly run through the content downloader since all the content is already available. The last step would to apply the tags per the chosen content pack would still run.